### PR TITLE
Feature: validate query method

### PR DESCRIPTION
**Pull request by calebmer**
I'm doing some work on `ardite-rest`, as I find useful methods I'll implement them. `validate_value` is going to suck to implement…

### CODE REVIEW

**Comment by calebmer** [comment]
This is not ready to be merged btw

File: src/driver.rs [hunk 1]
```diff
@@ -11,7 +11,9 @@ use query::Query;
 /// exist in the file system, and then return an instance initialized with the
 /// `connect` static trait function.
 pub fn get_driver<D: Driver>(url: &str) -> D {
-  // TODO: implement
+  // TODO: implement.
+  // Used to quiet warnings.
```
Comment (svmnotn) [inline src/driver.rs]: You know you can just quiet the warnings directly with a #[allow(...)]

File: src/lib.rs [hunk 2]
```diff
@@ -1,5 +1,8 @@
-#![allow(dead_code, unused_variables, unused_imports)]
+#[macro_use] extern crate lazy_static;
```
Comment (svmnotn) [inline src/lib.rs]: #[macro_use] on top of instead of same line is my preferred style.

File: src/driver.rs (hunk 1 above)
Comment (calebmer) [inline src/driver.rs]: That disables all of the warnings everywhere in the project, we don't want that.

Comment (calebmer) [inline src/driver.rs]: Wait, can we do local `#[allow]`s?

File: src/lib.rs (hunk 2 above)
Comment (calebmer) [inline src/lib.rs]: 👍

File: src/driver.rs (hunk 1 above)
Comment (svmnotn) [inline src/driver.rs]: Yup. The default is to be local. I had to put a `!` for it to apply to the whole module.

**Comment by svmnotn** [comment]
can we have an internal rule where if the module is, lets say over 200 lines, to represent too long (tests not included unless they are over ... idk what seems too long for tests? 300?) we split it into its own module?

Since that does seem to be what happened to `schema`.

**Comment by calebmer** [comment]
I don't think so. For javascript I'd split stuff up into lots of small files, but for all the codes I've seen, large files seem to be preferred in rust. I split up schema because it made sense taxonomically in my mind 😊

I'm actually going to be adding another giant multi-armed function `validate_value` to `schema.rs` soon so a line limit doesn't make a whole lot of sense.

Also, I'm repurposing this PR to just the addition of `validate_query` + refactors, merge whenever you feel comfortable with the PR.

**Comment by svmnotn** [comment]
I think it would be better to split this PR into two, one that does the bulk of the changes that you have so far, and then one the impls the `validate_query` method. 

Also just to note: Stop messing with the `.gitignore`, how many times do we have to have this discussion about you removing the windows' Desktop.ini file from it?

**Comment by calebmer** [comment]
I agree this probably should be two PRs, but I really don't want to do the git lifting work to split this PR into two 😊

**Comment by calebmer** [comment]
What's blocking getting this merged? I don't want to merge until you are totally on board.

File: src/schema.rs [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    match (self, query) {
+      // No schema describes these values, its the wild west. Go crazy query.
+      // `Schema::None` does not represent the absence of value, just the
+      // absence of validation.
+      (&Schema::None, _) => Ok(()),
+      (&Schema::Null, &Query::Value) => Ok(()),
+      (&Schema::Null, &Query::Object(_)) => Err(Error::validation("Cannot deeply query null.", NO_PRIMITIVE_HINT)),
+      (&Schema::Boolean, &Query::Value) => Ok(()),
+      (&Schema::Boolean, &Query::Object(_)) => Err(Error::validation("Cannot deeply query a boolean.", NO_PRIMITIVE_HINT)),
+      (&Schema::Number{..}, &Query::Value) => Ok(()),
```
Comment (svmnotn) [inline src/schema.rs:87]: This compiles?

Other than that feel free to merge.

Comment (calebmer) [inline src/schema.rs:87]: I don't know either, but the tests pass 👌

Out of curiosity what smells to you?

Comment (svmnotn) [inline src/schema.rs:87]: `Number{..}` To be specific the `..` inside the `{}`

Comment (calebmer) [inline src/schema.rs:87]: Yeah, Rust forces you to destructure all the things. `..` is the rest operator I guess and basically means ignore the stuff I didn't include.

**Pull request merged**